### PR TITLE
[marvell-armhf] Armada A385 soc support

### PR DESCRIPTION
**Pull request by antony-rheneus**
Add platform specific SOC kernel configuration based on CONFIGURED_PLATFORM variable passed from sonic-buildimage.

Marvell patches applied to Linux 4.19 are removed, 
Kernel driver for SFF, optoe Eeprom are enabled for Armhf architecture.
Due to a platform limitation the (compressed) Linux kernel size can only be 6 MB, so the Linux kernel configuration is shrinked to reduce the size.
Enable MTD parts and command line to access the nand flash

### CODE REVIEW

**Comment by antony-rheneus** [comment]
Duplicate of PR 172 https://github.com/Azure/sonic-linux-kernel/pull/172, 
Opened this to addess the review comments, which I could not proceed in the same PR 172 due to multiple merge conflicts during git rebase.

**Comment by lguohan** [comment]
@paulmenzel, can you take a look at this pr?

**Comment by antony-rheneus** [comment]
@paulmenzel , Would you please review the PR?

**Comment by paulmenzel** [comment]
Sorry, I thought, I replied in https://github.com/Azure/sonic-linux-kernel/pull/172 to keep discussion there.

**Comment by antony-rheneus** [comment]
> Sorry, I thought, I replied in #172 to keep discussion there.

Marvell armhf requires configuration from "sonic-linux-kernel/linux-4.19.118/arch/arm/configs/mvebu_v7_defconfig" however it is not so when debian prepares the .config.

**Comment by carl-nokia** [comment]
Paul - what is missing or required from Marvell/Nokia teams to merge this PR? -- we need to get this merged ASAP please

**Comment by paulmenzel** [comment]
> Paul - what is missing or required from Marvell/Nokia teams to merge this PR? -- we need to get this merged ASAP please

Please read the discussion in merge/pull request https://github.com/Azure/sonic-linux-kernel/pull/172.

**Comment by carl-nokia** [comment]
Please forgive my lack of clarity on this PR, but I did read PR 172 before asking you for help on this PR 176.   I just re-read it again.  I want to help resolve whatever is needed ASAP.    
        The best I can tell is that you are still waiting for the PR description and  23 commit messages to follow **"How to Write a Git Commit Message"**  best practices properly.

Is my understanding correct?  ... or are you also looking for a rebase to better define the list of commits for review?

**Comment by paulmenzel** [comment]
Yes, your understanding is correct.

Then for example, it might become clear, why you have to add certain configs to the inclusion file, and it’s opposite configuration to the exclusion file at the same time (`CONFIG_KERNEL_GZIP=y` vs. `CONFIG_KERNEL_XZ`. Only one should be necessary, as it’s a choice.

**Comment by paulmenzel** [comment]
Having to merge/pull requests is now even more confusing, but so be it. For example https://github.com/Azure/sonic-linux-kernel/pull/176/commits/9976144f33fea5a946c5ca8f30aeae5081a7e609:

>  [kernel] [marvell-armhf]
>
> Set cpu frequenc to 100HZ as per armada cpu
> Enable periodic timer and not dynamic for better performance of a385 soc

I asked several times to follow the guide on how to write commit messages. The summary belongs in the first line for example, and that’s what it at the very beginning and also well seen in the [Linux kernel](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git) and [git project repository](https://git.kernel.org/pub/scm/git/git.git/).

Why the kernel tag? It’s the *sonic-linux-kernel* repository. Maybe:

> marvell-armhf: armada: Set CPU frequency to 100HZ and use periodic timer
>
>  Enable periodic timer and not dynamic for better performance of a385 soc.
>
> [Add explanation here, why 100 HZ increases the performance, how you tested it, and what the performance increase is.]

**Comment by antony-rheneus** [comment]
@paulmenzel, @lguohan, 
 We cleaned up all the commit messages to make them more descriptive, following established github guidelines. The "Why" of all the commits were carefully added. Please review and let us know if any additional changes are required.

**Comment by paulmenzel** [comment]
Thank you, much better!

Some nits:

1.  Why is GZIP in the inclusions not enough? Does the script have an error, so that KERNEL_XZ has to be added to the excludes?
2.  For the kernel image size reduction commit, can you please document what the old and new size is?
3.  not appened → not appended in the last commit message.

**Comment by antony-rheneus** [comment]
> Thank you, much better!
> 
> Some nits:
> 
> 1. Why is GZIP in the inclusions not enough? Does the script have an error, so that KERNEL_XZ has to be added to the excludes?
> 2. For the kernel image size reduction commit, can you please document what the old and new size is?
> 3. not appened → not appended in the last commit message.


@paulmenzel 
Please review the update commits in the PR, Let us know if any additional changes required. 

#1 Since only one compression method can be selected,  added config to exclude XZ as it came as part of default config. The conf script didn't report any error. 

#2 Documented the file size in the commit description as suggested, please check and confirm if that is fine.
Before size: 6197760
After  size: 5192480

#3 Corrected the typo "appened" in commit description.

**Comment by lguohan** [comment]
@antony-rheneus, it looks like this gz v.s xz issue has been brought up multiple times, and looks like xz is prefered over gz. since armhf maybe used for other sku in the future, can you change to xz instead of gz, so that in the future we won't need to discuss this option again.

**Comment by antony-rheneus** [comment]
> @antony-rheneus, it looks like this gz v.s xz issue has been brought up multiple times, and looks like xz is prefered over gz. since armhf maybe used for other sku in the future, can you change to xz instead of gz, so that in the future we won't need to discuss this option again.

@lguohan, @paulmenzel , 
As requested, I have revert the GZIP compression, and tested the same in our platform with XZ compression method.
Also I hase rebased to master to resolve the conflicts with patch/series.
Please do review the recent changes.

**Review by lguohan — APPROVED** [review]

**Comment by lguohan** [comment]
lgtm

File: patch/kconfig-inclusions [hunk 1]
```diff
@@ -58,5 +58,19 @@ CONFIG_MAX1363=m
...
 [armhf]
+CONFIG_EEPROM_SFF_8436=m
+CONFIG_EEPROM_OPTOE=m
+CONFIG_I2C_MUX_GPIO=m
 
-
+[marvell-armhf]
+CONFIG_HZ_PERIODIC=y
+CONFIG_HZ_100=y
+CONFIG_HZ=100
```
Comment (paulmenzel) [inline patch/kconfig-inclusions:68]: 1.  Your [commit message](https://github.com/Azure/sonic-linux-kernel/pull/176/commits/3e00dbc2aaf8db8fe7b2450a0674f1529c06a01a) says, it’s in `linux-4.19.118/arch/arm/configs/mvebu_v7_defconfig`. But I do not find it there, and it’s also not in [Linux 5.10](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm/configs/mvebu_v7_defconfig?h=v5.10).
 
2.  You should only need `CONFIG_HZ_100=y`. `CONFIG_HZ=100` will be set by Linux’ Kconfig build system as a consequence.

File: patch/kconfig-exclusions [hunk 2]
```diff
@@ -10,3 +10,35 @@ CONFIG_MLXSW_PCI
 [arm64]
 
 [armhf]
+
+[marvell-armhf]
+CONFIG_NO_HZ_IDLE
+CONFIG_HZ_250
+CONFIG_SLUB
+CONFIG_DEBUG_SLAB
```
Comment (paulmenzel) [inline patch/kconfig-exclusions:18]: Why is that added here? The config defaults to n, doesn’t it?

File: patch/kconfig-exclusions [hunk 3]
```diff
@@ -10,3 +10,35 @@ CONFIG_MLXSW_PCI
 [arm64]
 
 [armhf]
+
+[marvell-armhf]
+CONFIG_NO_HZ_IDLE
+CONFIG_HZ_250
```
Comment (paulmenzel) [inline patch/kconfig-exclusions:16]: I am sorry, there is still a big misunderstanding going on here. I try to make my point here. If you want 100 Hz as in commit https://github.com/Azure/sonic-linux-kernel/pull/176/commits/3e00dbc2aaf8db8fe7b2450a0674f1529c06a01a you should only need to add it to the *inclusion* file, and the Linux’ Kconfig system will disable the other options (because it’s a choice). Having the “choice option” also in the *exclusion* file should be unnecessary, and only add maintenance effort (if you want to change to 250 Hz, you have to edit several files) and confusing. So, please only add `CONFIG_HZ_100=y`

**Review by paulmenzel — COMMENTED** [review]
Sorry, I feel like I am missing something, that I have all these questions, nobody else has during review. Also, I am now spent over one hour with the merge/pull request, which actually tries to do a simple thing. I know you spent more time on it, but I actually do not get paid for this, and only stumbled upon in, while looking at the “project status” of SONiC.

**Comment by carl-nokia** [comment]
Paul - I will set up a meeting with you - this merge is blocking many people - lets get on a call please to answer your questions - we need to wrap this up please. I will send you invite and cc Guohan now.

**Comment by paulmenzel** [comment]
> Paul - I will set up a meeting with you - this merge is blocking many people - lets get on a call please to answer your questions - we need to wrap this up please. I will send you invite and cc Guohan now.

Thanks. It would be nice though to get some written reply to my comments beforehand.

File: patch/kconfig-inclusions [hunk 4]
```diff
@@ -58,5 +58,19 @@ CONFIG_MAX1363=m
...
+CONFIG_SLAB=y
+CONFIG_COMPAT_BRK=y
+CONFIG_DRM_PANEL_ORIENTATION_QUIRKS=m
```
[the remark this answers is not in the document]
Comment (antony-rheneus) [reply patch/kconfig-inclusions]: DRM is not needed, also the changes were taken from Marvell private version of mvebu_v7_defconfig.
Hence removing it to be in sync with the upstreamed vebu_v7_defconfig

File: patch/kconfig-inclusions (hunk 1 above)
Comment (antony-rheneus) [reply patch/kconfig-inclusions]: 1. The changes were taken from Marvell private version of mvebu_v7_defconfig.
Hence removing it to be in sync with the upstreamed mvebu_v7_defconfig

2. Reverted the change to avoid conflicts with the upstreamed configuration

File: patch/kconfig-exclusions (hunk 3 above)
Comment (antony-rheneus) [reply patch/kconfig-exclusions]: Reverted the changes and we made config to be inline with the generated configuration.

**Comment by antony-rheneus** [comment]
@paulmenzel , @lguohan ,
I have reverted the configurations to be in sync with the upstreamed version for mvebu_def_config. Let us know if you have any comments.

**Comment by antony-rheneus** [comment]
Below are the recent commits under review.

```
$ git log --oneline -n 7
bd6a99e armhf: Update linux debian image file name
873ee9f Marvell-armhf: Disable uefi firmware for marvell
305dc69 Marvell-armhf: Reduce kernel image size for u-boot
b6ee695 Marvell-armhf: Enable optoe and i2c mux
b78363e Marvell-armhf: armada: Set gpio and mtd platform specific configuration
65f6779 Marvell-armhf: Enable platform specific kconfig
a3d548c Marvell-armhf: Remove obsolete patches applied for marvell arm

```

**Comment by radha-danda** [comment]
> > Paul - I will set up a meeting with you - this merge is blocking many people - lets get on a call please to answer your questions - we need to wrap this up please. I will send you invite and cc Guohan now.
> 
> Thanks. It would be nice though to get some written reply to my comments beforehand.

@paulmenzel, Antony addressed your comment and pushed the changes. Can you please review and let us know if you have any further comments

**Comment by paulmenzel** [comment]
Thank you for the last changes. In commit https://github.com/Azure/sonic-linux-kernel/pull/176/commits/305dc691085212b4c5f223b66bcfa420dc9f85b1 you cite the numbers below:

> Testing:
>  - Check kernel boots without any memory overlapping errors
>  - kernel binary size before this commit : 6197760 bytes
>  - kernel binary size after this commit  : 5192480 bytes

Is that a hardware limitation? Or a U-Boot issue? Do you have an upstream bug report for that?

So as talked about in the conference, it would be great, if you documented the maximum allowed size, and what compression algorithm was used (if it actually plays a role). The merge/pull request description would be something like (just a made-up suggestion):

> Add platform specific SOC kernel configuration.
>
> Marvell patches applied to Linux 4.19 are removed, and due to a platform limitation the (compressed) Linux kernel size can only be 6 MB, so the Linux kernel configuration is shrinked to reduce the size.

**Comment by antony-rheneus** [comment]
> Thank you for the last changes. In commit [305dc69](https://github.com/Azure/sonic-linux-kernel/commit/305dc691085212b4c5f223b66bcfa420dc9f85b1) you cite the numbers below:
> 
> > Testing:
> > 
> > * Check kernel boots without any memory overlapping errors
> > * kernel binary size before this commit : 6197760 bytes
> > * kernel binary size after this commit  : 5192480 bytes
> 
> Is that a hardware limitation? Or a U-Boot issue? Do you have an upstream bug report for that?
> 
> So as talked about in the conference, it would be great, if you documented the maximum allowed size, and what compression algorithm was used (if it actually plays a role). The merge/pull request description would be something like (just a made-up suggestion):
> 
> > Add platform specific SOC kernel configuration.
> > Marvell patches applied to Linux 4.19 are removed, and due to a platform limitation the (compressed) Linux kernel size can only be 6 MB, so the Linux kernel configuration is shrinked to reduce the size.

Size constraint is hardware specific limitation for u-boot based platforms.

**Comment by antony-rheneus** [comment]
@paulmenzel , I have updated the comment description, and also the PR description. Would you please review?

**Comment by paulmenzel** [comment]
> Size constraint is hardware specific limitation for u-boot based platforms.

Do you have a reference for that?

**Comment by paulmenzel** [comment]
Nice, looks good.

Just to be sure: I made up the number in “Linux kernel size can only be 6 MB,”. That is correct.

Also, just to be sure:

> Testing:
>  - Check kernel boots without any memory overlapping errors
>  - kernel binary size before this commit : 6197760 bytes
>  - kernel binary size after this commit  : 3600624 bytes

The 6.2 MB is really XZ compressed image or was it with GZIP? Because if you try XZ which results in smaller size, maybe you can use the default kernel configuration.

**Comment by paulmenzel** [comment]
Just to be clear, although I am interested in the answers of the questions above, I do not want to hold anything up. So, if @lguohan wants to accept this, please go ahead.

**Comment by lguohan** [comment]
@paulmenzel , thanks for the reviewing. the pr is good. @antony-rheneus , please try to answer the questions, I am also curious to learn.

**Comment by paulmenzel** [comment]
@antony-rheneus, it’d be great if you replied to my questions.

**Comment by antony-rheneus** [comment]
> @antony-rheneus, it’d be great if you replied to my questions.

@paulmenzel , sure, I will test the kernel with the changes and reply back.

**Comment by antony-rheneus** [comment]
> > @antony-rheneus, it’d be great if you replied to my questions.
> 
> @paulmenzel , sure, I will test the kernel with the changes and reply back.

@paulmenzel , 
After removing all the SOCs define in "exclusion" file, now the image is reduced to 4.2MB with xz compression
```
ls -l linux-4.19.118/debian/build/build_armhf_none_armmp/arch/arm/boot/zImage
-rwxr-xr-x 1 root root 4260352 Dec 21 23:10 linux-4.19.118/debian/build/build_armhf_none_armmp/arch/arm/boot/zImage

```
However, the kernel hangs, so there is some conflicting CONFIG variable when all SOCs are defined. I will reuire some time to debug this to narrow down the invalid configuration.
```

Starting kernel ...

<<<HUNG>>>
```

**Comment by paulmenzel** [comment]
Over three months have passed. Were you able to fix the issue?

**Comment by antony-rheneus** [comment]
> Over three months have passed. Were you able to fix the issue?

Not yet, we will get back.

**Comment by paulmenzel** [comment]
@carl-nokia, could you (or the responsible project manager) please allocate time for @antony-rheneus or some other developer/engineer to work on this?

**Comment by paulmenzel** [comment]
Over a year has passed, and the *promised* work, convincing the reviewers to merge the changes, has not happened yet.

**Comment by antony-rheneus** [comment]
@paulmenzel, 
Redirecting the request to the Project Manager (@radha-danda)

**Comment by radha-danda** [comment]
@paulmenzel, will sometime next week and get back to you
